src/fw-update/fw-update-factory.cpp

```diff
@@ -11,33 +11,100 @@
 #include "l500/l500-private.h"
 #include "l500/l500-fw-update-device.h"
 
+#define FW_UPDATE_INTERFACE_NUMBER 0
+#define DEFAULT_TIMEOUT 100
+
 namespace librealsense
 {
-    int get_product_line(uint16_t pid, platform::usb_class cls)
+    bool is_l500_recovery(platform::rs_usb_device usb)
     {
-        if (SR300_RECOVERY == pid && platform::RS2_USB_CLASS_VENDOR_SPECIFIC == cls)
-            return RS2_PRODUCT_LINE_SR300;
-        if(ds::RS_RECOVERY_PID == pid || ds::RS_USB2_RECOVERY_PID == pid)
-            return RS2_PRODUCT_LINE_D400;
-        if (L500_RECOVERY_PID == pid)
-            return RS2_PRODUCT_LINE_L500;
-        return 0;
+        dfu_fw_status_payload payload;
+
+        if (auto messenger = usb->open(FW_UPDATE_INTERFACE_NUMBER))
+        {
+            // Step 1 - Detach device - mandatory before communicating with the DFU
+            {
+                auto timeout = 1000;
+                uint32_t transferred = 0;
+                auto res = messenger->control_transfer(0x21, RS2_DFU_DETACH, timeout, 0, NULL, 0, transferred, timeout);
+                if (res != platform::RS2_USB_STATUS_SUCCESS)
+                    LOG_WARNING("DFU - failed to detach device");
+            }
+
+            // Step 2 - Verify device entered DFU idle state and ready to communicate with the host
+            {
+                uint8_t state = RS2_DFU_STATE_DFU_ERROR;
+                uint32_t transferred = 0;
+                auto res = messenger->control_transfer(0xa1, RS2_DFU_GET_STATE, 0, 0, &state, 1, transferred, DEFAULT_TIMEOUT);
+                if (res != platform::RS2_USB_STATUS_SUCCESS)
+                    throw std::runtime_error("Failed to get DFU device state!");
+
+                if (state != RS2_DFU_STATE_DFU_IDLE)
+                    throw std::runtime_error("DFU detach failed!");
+            }
+            // Step 3 - Query DFU information
+            {
+                uint32_t transferred = 0;
+                auto sts = messenger->control_transfer(0xa1, RS2_DFU_UPLOAD, 0, 0, reinterpret_cast<uint8_t*>(&payload), sizeof(payload), transferred, DEFAULT_TIMEOUT);
```

Comment:
Where do we reattach?

I don't know if I like this... a purely informational function does a detach??

Reply:
According to DFU protocol spec there is no reattach command.
Only a USB reset will trigger an application state.
While on DFU (HWM command) you get into a DFU disabled state.
A detach command is necessary for communicating with the DFU device.
You exit DFU state by USB reset only.
So 
1. detach command on DFU mode is harmless to other processes.
2. It is necessary for communicating with it / reading the DFU information.

We got to this function because the device is on DFU mode already.
I can splot the stages into functions.. did not wantes to create duplicated functions.
Any other suggestions?